### DISCUSSION TITLE
How to do inference on gray scale images?

### DISCUSSION THREAD

**Discussion by vinayakarannil** [facebookresearch/detectron2#3182] [Q&A]
I have trained the model using 'L' mode, but inference does not support gray scale mode. How would i do inference on my test data? I can run in BGR or RGB mode. But since training was done on gray scled images, the accuracy might be impacted if i change the color mode in inference right?

**Comment by johnanthonyjose** [comment]
Currently, I am uncertain what you mean by 'L' mode in detectron2. 

What do you mean by "you are able to train in grayscale but cannot perform inference in grayscale"? If detectron2 is able to support grayscale training, then it should also mean that you should be able to support grayscale inference too.

Have you checked your dimension in your inference? Pytorch requires that the tensor dimension is at least four dimension: [BxCXHxW] where B = batch size, C = number of channel, H = height in pixels, W = width in pixels.

If it's grayscale then C dimension is equal to 1

**Reply by vinayakarannil** [reply]
My dear Brother,

I would really appreciate, if you do some background study before commenting on such posts. My apologies, if my question was confusing.

By 'L' mode, i mean the INPUT.FORMAT option which we set in detectron2 config . Detectron2 internally uses PIL library for image handling. So 'L' mode in pillow means grayscale(https://pillow.readthedocs.io/en/stable/handbook/concepts.html). 

 During training,  detectron2 accepts INPUT.FORMAT as 'L' , along with other modes like 'RGB','BGR', and so on. 
While on inference if you provide INPUT.FORMAT as 'L', you will run into assertion error, because for inference it accepts only 'RGB' or 'BGR' modes. There is a strict assertion kept here https://github.com/facebookresearch/detectron2/blob/5d686809a2782e424f81bb5154e140bc0bb7db6f/detectron2/engine/defaults.py#L298

Nevertheless, i can send my gray scale image with INPUT.FORMAT= 'RGB', by expanding the dim before sending to the model. But i wanted to understand, why such an assertion was kept only for inference.
